docs/ProofOptimization/ProofOptimization.md

```diff
@@ -0,0 +1,599 @@
+
+## Non-linear arithmetic
+
+Non-linear arithmetic in Dafny generally takes the form of multiplication operations where both sides are variables or any instance of the division or modulus operations. Proving arbitrary facts about operations of this form is undecidable, so the solver must use incomplete heuristics, and frequently gives up.
+
+Fortunately, Dafny gives you the power to prove complex facts about non-linear arithmetic by building on slightly less complex facts, and so on, down to the point where the solver can prove the smallest building blocks automatically.
+
+There are a number of examples in the Dafny [`libraries`](https://github.com/dafny-lang/libraries) that do just this. For example, consider the following sequence of lemmas that builds up from basic properties about multiplication all the way to basic properties about exponentiation.
+
+<!-- %check-verify -->
+```dafny
+lemma LemmaMulBasics(x: int)
+  ensures 0 * x == 0
+  ensures x * 0 == 0
+  ensures 1 * x == x
+  ensures x * 1 == x
+{
+}
+
+opaque function Pow(b: int, e: nat): int
+  decreases e
+{
+  if e == 0 then
+    1
+  else
+    b * Pow(b, e - 1)
+}
+
+lemma LemmaPow0(b: int)
+  ensures Pow(b, 0) == 1
+{
+  reveal Pow();
+}
+
+lemma LemmaPow1(b: int)
+  ensures Pow(b, 1) == b
+{
+  calc {
+    Pow(b, 1);
+    { reveal Pow(); }
+    b * Pow(b, 0);
+    { LemmaPow0(b); }
+    b * 1;
+    { LemmaMulBasics(b); }
+    b;
+  }
+}
+```
+
+All of these lemmas can be proved and used with the `--disable-nonlinear-arithmetic` flag enabled. These lemmas also demonstrate several techniques that we'll describe in more detail later: lots of small lemmas, opaque functions, and [`calc`](../DafnyRef/DafnyRef#sec-calc-statement) statements.
+
+If you have a difficult non-linear fact to prove, and proving it from the ground up is infeasible, it is possible, as a last resort, to adjust how Z3 attempts to prove arithmetic formulas. It contains a parameter, `smt.arith.solver`, that allows you to choose between one of six different arithmetic decision procedures. Dafny has historically used solver "2", and continues to do so by default. This peforms most effectively on the corpus of examples that we have experimented with. However, solver "6", which is the default for Z3 when not being invoked from Dafny, performs better at non-linear arithmetic. Currently, it's possible to choose this solver using the Dafny flag `--boogie /proverOpt:O:smt.arith.solver=6` (or just `/proverOpt:O:smt.arith.solver=6`, when not using top-level commands).
+
+## Relationships between integers and bit vectors
+
+Dafny contains two distinct ways to describe what might be stored in a machine word in other programming languages.
+
+* The `int` type denotes the unbounded mathematical integers.
```

Comment:
Perhaps mention that in practice you'd use a bounded `newtype` and not `int` directly, especially since then you really are directly equivalent with the bit vector option.